mono/metadata/threads.c

```diff
@@ -1650,40 +1655,60 @@ mono_thread_set_name_internal (MonoInternalThread *this_obj, MonoString *name, g
 		mono_error_set_invalid_operation (error, "Thread.Name can only be set once.");
 		return;
 	}
-	if (this_obj->name) {
-		g_free (this_obj->name);
-		this_obj->name_len = 0;
-	}
-	if (name) {
-		this_obj->name = g_memdup (mono_string_chars (name), mono_string_length (name) * sizeof (gunichar2));
-		this_obj->name_len = mono_string_length (name);
 
+	g_free (this_obj->name);
+	this_obj->name = 0;
+	this_obj->name_len = 0;
+
+	if (length > 0) {
+		this_obj->name = utf16 ? (gunichar2*)g_memdup (utf16, length * sizeof (gunichar2))
+							   : g_utf8_to_utf16 (utf8, length, NULL, NULL, NULL);
```

Comment:
Do we know that `g_utf8_to_utf16` won't fail? Shouldn't we capture its error message if it does?

Reply:
Ok how about? https://github.com/jaykrell/mono/commit/d2a3088cd548cd832aea0acd67990a1b605dc191

Reply:
Ok how about ? https://github.com/jaykrell/mono/commit/d815c874cd35c0cc369d8fd795bca6ba076d8064